framework/include/auxkernels/MaterialAuxBase.h

```diff
@@ -50,26 +68,94 @@ InputParameters
 MaterialAuxBaseTempl<T, is_ad, RT>::validParams()
 {
   InputParameters params = AuxKernelTempl<RT>::validParams();
-  params.addRequiredParam<MaterialPropertyName>("property", "The scalar material property name");
+  params.addParam<MaterialPropertyName>(
+      "property", "The scalar material property name (set either this or `functor`).");
+  // functors do not support all types
+  if constexpr (!std::is_same_v<MaterialAuxFunctorType, void>)
+    params.addParam<MooseFunctorName>(
+        "functor", "The scalar material property name (set either this or `property`).");
   params.addParam<Real>(
       "factor", 1, "The factor by which to multiply your material property for visualization");
   params.addParam<RT>("offset", 0, "The offset to add to your material property for visualization");
+
+  params.addParam<unsigned int>(
+      "selected_qp",
+      "Evaluate the material property at a specified quadrature point. This only needs "
+      "to be used if you are interested in a particular quadrature point in each element. "
+      "Otherwise do not include this parameter in your input file.");
+  params.addParamNamesToGroup("selected_qp", "Advanced");
+
   return params;
 }
 
 template <typename T, bool is_ad, typename RT>
 MaterialAuxBaseTempl<T, is_ad, RT>::MaterialAuxBaseTempl(const InputParameters & parameters)
   : AuxKernelTempl<RT>(parameters),
-    _prop(this->template getGenericMaterialProperty<T, is_ad>("property")),
+    _property(this->isParamValid("property")
+                  ? &this->template getGenericMaterialProperty<T, is_ad>("property")
+                  : nullptr),
+    _functor(this->isParamValid("functor")
+                 ? [this]() {
+                    if constexpr (!std::is_same_v<MaterialAuxFunctorType, void>)
+                      return &this->template getFunctor<GenericType<T, is_ad>>("functor");
+                    else
+                    {
+                      libmesh_ignore(this);
+                      return nullptr;
+                    }
+                 }()
+                 : nullptr),
+    _selected_qp(this->isParamValid("selected_qp") ? this->template getParam<unsigned int>("selected_qp") : libMesh::invalid_uint),
     _factor(this->template getParam<Real>("factor")),
     _offset(this->template getParam<RT>("offset"))
 {
+  if (!_property == !_functor)
+    mooseError("Specify either a `property` or a `functor` parameter.");
+  if (_functor && _selected_qp != libMesh::invalid_uint)
+    this->paramError("selected_qp",
+                     "Selective quadrature point evaluation is not implemented for functors.");
 }
```

Comment:
Can you also check that if `_selected_qp` is provided that the aux kernel is not nodal? Technically it could work if the user provides a qp of 0 but I still think it's logically silly

Reply:
This has no support for nodal evaluation yet. The part that evaluates the functor value would have to be updated. I can add that.

Reply:
You say there is no support, but I don't see an error for if the kernel is nodal?

Reply:
Working on it... 
However there isn't even a `selected_qp` parameter for the Functor version of these classes, and you automatically get an error when getting material properties in a nodal AuxKernel. I'm adding support for nodal functor evaluation.

Reply:
Added